pypeit/core/skysub.py

```diff
@@ -1210,35 +1222,62 @@ def ech_local_skysub_extract(sciimg, sciivar, fullmask, tilts, waveimg,
     ivarmodel = np.copy(sciivar)
     sobjs = sobjs.copy()
 
-    norders = order_vec.size
-    slit_vec = np.arange(norders)
+    # Identify the unique SLITIDs and orders in the sobjs object
+    slitids = np.unique(sobjs.SLITID) # This will also sort the slitids
+    norders = (np.unique(sobjs.ECH_ORDER)).size
 
     # Find the spat IDs
-    gdslit_spat = np.unique(slitmask[slitmask >= 0]).astype(int)  # Unique sorts
-    #if gdslit_spat.size != norders:
-    #    msgs.error("You have not dealt with masked orders properly")
-
-    #if (np.sum(sobjs.sign > 0) % norders) == 0:
-    #    nobjs = int((np.sum(sobjs.sign > 0)/norders))
-    #else:
-    #    msgs.error('Number of specobjs in sobjs is not an integer multiple of the number or ordres!')
-
-    # Set bad obj to -nan
+    # TODO JFH Relaxing this requirement for now since it would require that we regenerate the slitmask image.
+    # each time slits are masked, which is probably not necessary. Do we want to enforce this on the slitmask image?
+
+    #gdslit_spat = np.unique(slitmask[slitmask >= 0]).astype(int)  # Unique sorts
+
+    #if not np.array_equal(gdslit_spat, slitids):
+    #    msgs.error('The slitmask image and the sobjs object do not have the same number of good orders! '
+    #               'There is a problem with the object/slitmask masking. This routine requires that all '
+    #               'masking is performed in the calling routine.')
+    if norders != len(slitids):
+        msgs.error('The number of orders in the sobjs object does not match the number of good slits in the '
+                   'slitmask image! There is a problem with the object/slitmask masking. This routine '
+                   'requires that all masking is performed in the calling routine.')
+
+    # Check that the slit edges are masked consistent with the number of orders and the number of unique spatids
+    nleft = left.shape[1]
+    nrigh = right.shape[1]
+    if (nleft != nrigh) or norders != nleft or norders != nrigh:
+        msgs.error('The number of left and right edges must be the same as the number of orders. '
+                   'There is likely a problem with your masking')
+
+    # Now assign the order_sn, and generate an order_vec aligned with the slitids
     uni_objid = np.unique(sobjs[sobjs.sign > 0].ECH_OBJID)
     nobjs = len(uni_objid)
-    order_snr = np.zeros((norders, nobjs))
-    order_snr_gpm = np.ones_like(order_snr) 
-    for iord in range(norders):
+    order_snr = np.full((norders, nobjs), np.nan)
```

Comment:
I think we can agree to disagree. I hate numpy masked arrays and find them cumbersome and buggy. Furthermore, the correct value is actually nan here, since an unassigned value is not a number.